### DISCUSSION TITLE
Share your experience with phpIP

### DISCUSSION THREAD

**Discussion by jjdejong** [jjdejong/phpip#127] [Q&A]
Ask questions, suggest new features or simply let us know you are a user. 
You need to sign in with a GitHub account to post comments and subscribe to the discussion. 
Do not post issues and bug reports here but in the dedicated [Issues](https://github.com/jjdejong/phpip/issues) tab).

**Comment by MichelPellegri** [comment]
Hello, I have a question concerning phpip.
How could I do to add "Cl. Ref" on a file ?
Thank you for your help !

**Reply by jjdejong** [reply]
On the actor role header there is an "Edit" button. Click that and add the client reference in the corresponding field.
<img width="318" alt="image" src="https://github.com/jjdejong/phpip/assets/1052566/55a9ba24-2301-496c-ac78-0508becdd198">

**Comment by clairey69** [comment]
Hi, 
I'm using outlook for sending emails. I have created a few email templates in phpIP (in txt), but if the text is too long, then the "send" button does not generate any email. I never had any issues with the length of my emails in Outlook. Is there a maximum number of characters for phpIP email's templates ?

Thank you !

**Reply by jjdejong** [reply]
The functionality generates a `mailto:` URI that your navigator forwards to your "default email client". The interpretation of the URI depends on your navigator and your email client, which may impose a limit on the URI length, so there is nothing that can be done about it within phpIP. Try using another email client or another navigator (Firefox is what we use for the development). Note that Thunderbird is very "advanced" in interpreting `mailto:` URIs, since it even accepts email bodies in HTML.

**Reply by papoteur-mga** [reply]
To complete the answer, I tested the same template of about 5000 characters. It works from Firefox, but not from Chrome. The latter seems to have a limit around 2000 characters, according to this discussion [this discussion](https://stackoverflow.com/questions/4473959/effective-maximum-mailto-body-lengths).

**Reply by clairey69** [reply]
Thank you both ! I confirm that if I use Firefox the problem disappears :)

**Comment by clairey69** [comment]
Hi ! Is it possible to search matters by country ? Thank you !

**Reply by jjdejong** [reply]
Sure, either add `country=CC` to the URL as a parameter (preceded by `?` if it's the only parameter or by `&` otherwise) or type `%CC` in the `Ref` filter field (`%` is the wild character).

**Reply by clairey69** [reply]
thank you, it works like a charm !

**Comment by iplaw67** [comment]
Hi,
Should I ask questions about the installation instructions here, or file a bug report directly ?

For background, I tried installing phpip onto a Raspberry Pi4 with 8Mb RAM with Ubuntu 22.04 Desktop for that Arm architecture. 
The installation sequence fails initially at the mysql setup, when the following DDL instruction, as provided on the wiki is executed:

"CREATE USER phpip@localhost IDENTIFIED BY 'phpip';"

When I execute the above instruction, I get an error message that it doesn't respect the default encryption required by the mysql server version.

Should I file a bug against the documentation ?

**Comment by iplaw67** [comment]
This might have already been discussed, but is there any function or tool within the phpIP UI that allows you to import data from a spreadsheet or CSV file, or does that have to be handled directly via inserts into the corresponding table on the mysql server ?

**Reply by iplaw67** [reply]
In a similar vein, is there a backup or export tool within the phpIP UI or would I need to do a mysql dump ?

**Reply by papoteur-mga** [reply]
The answer is "no" for both questions.
The import has to be managed by direct work in database.
All data are stored in the database, thus a mysql dump will be OK.

**Reply by iplaw67** [reply]
Thanks !

**Reply by jjdejong** [reply]
> In a similar vein, is there a backup or export tool within the phpIP UI or would I need to do a mysql dump ?

There is an export tool in the UI (Matters -> All, "Export" button), but limited to barely more than the displayed data - useful for creating portfolio statuses for clients.

**Comment by clairey69** [comment]
Hello,
I'd like to add the Client's address in an email template. I have tried {{ $matter->actors->where('role_code', 'CLI')->pluck('address')->join(', ') }} but it doesn't seem to work.
Can you please help ?
Thank you !

**Reply by jjdejong** [reply]
`$matter->actors->where('role_code', 'CLI')->first()->actor->address`
We use `first()`because there should only be one client (your syntax assumes there can be multiple clients).
But there is a specific property for the (single) client:
`$matter->client->actor->address`

**Reply by clairey69** [reply]
thank you !!

**Comment by clairey69** [comment]
Hi, 
I just found the "recurring" checkbox in rule informations : 
![image](https://github.com/jjdejong/phpip/assets/163107592/fdc494c1-3e41-47e7-be05-5450e846daa4)
What does it do ? Can I use it to generate the same task every 3 years for example ?
Thanks !

**Reply by jjdejong** [reply]
The name is misleading, because it was initially intended for another purpose that could not be implemented. The fields of all tables are described in the [Table Structure Wiki](https://github.com/jjdejong/phpip/wiki/Tables#task_rules).

**Comment by clairey69** [comment]
would it be possible to implement recurring tasks ? or did you encounter technical issues ? Because that would be extremely useful for trademarks.

**Reply by jjdejong** [reply]
The purpose _was_ for trademark renewals or any other multi-annual tasks. There is a technical barrier in MySQL for the part "_when the current task is closed, the next one is calculated_" -  MySQL triggers cannot modify the same table they are run from. A similar functionality could be developed outside MySQL, but it would break the consistency with the other functions and make the code difficult to maintain, so I didn't get around to it.

An alternative is to create a workflow, where you enter a "Renewal Paid" event that clears the current renewal and triggers the next, i.e. you set a first rule that is triggered by the "Renewal Paid" event and clears the pending renewal, and you set a second rule also triggered by the "Renewal Paid" event, that creates the next renewal task.

**Reply by clairey69** [reply]
many thanks for the detailed answer ! I will try creating a workflow and see if it works.

**Comment by clairey69** [comment]
Hello !
When I go to "email templates", they are displayed alphabetically. But when I send an email from a matter, the list of templates are displayed by date of creation (which means they are all mixed up). Would it be possible to set an alphabetical order when we want to send an email ?
Thank you !

**Reply by papoteur-mga** [reply]
Hi Claire,
Do you mean by this way, the button at right?
![image](https://github.com/jjdejong/phpip/assets/5154580/599dba51-8814-4d44-bb52-cf5dbadeec96)

**Reply by clairey69** [reply]
yes !

**Reply by jjdejong** [reply]
I created an issue #133.

**Reply by papoteur-mga** [reply]
@clairey69 I invite you to see the last alinea of this: https://github.com/jjdejong/phpip/wiki/Templates-(email-and-documents)#classes

You can organize your email templates by classes, then link these classes to events or tasks. The advantage is that the email can then be populated with dates coming from said event field or task field. Furthermore, the choice will be presented with only pertinent templates.

**Comment by iplaw67** [comment]
1) What would be required to allow for a differentiation of trademark G&S per class ?

2) Can the UI be changed to at least allow for a UI control that allows one to visualise the goods & services associated with a class selected from e.g. a dropdown menu ?

I'm thinking something like the mockup below:
![MockupUI-GS](https://github.com/user-attachments/assets/3208c11d-533b-43f1-9aa9-50365117d2ae)

**Reply by jjdejong** [reply]
phpIP has been coded by patent attorneys for handling patents, essentially. Of course, some attention was brought to handling trademarks and other matters. A focus was also to not duplicate information available elsewhere. Hence, things like importing into phpIP tables the information on TM classes is not something that was foreseen in the design. You can manually add the class information as "classifiers", which is what trademark users do currently. In those classifiers, you can add links to webpages that provide further information on the classes. Otherwise the addition of the feature would require substantial development.

**Reply by iplaw67** [reply]
When you say that you can add links to webpages, presumably, you mean a variety of webpages stored outside of phpip that would contain the specific class information for that particular case? I understand that the idea was to avoid duplication, but that doesn't cover the situation where a same basic trademark, e.g. international TM, may have different G&S in different countries depending on whether the national office has objected to the list of G&S and required amendment to conform to its national practice.

**Reply by jjdejong** [reply]
Right, the "classifiers" section is shared throughout the family members, so you can't add different classes in different countries. What TM users do is that they create multiple class lists for multiple countries. A class list is free text, so you can indicate country codes. You need to ask @clairey69, who uses phpIP for trademarks.
![image](https://github.com/user-attachments/assets/525ca2a5-74a0-4eb0-8c91-fa9f44ff65b0)

**Reply by jjdejong** [reply]
I forgot the first question. In any "classifier", you can insert a URL that will turn the classifier value into a link to that URL. So there you would insert a URL of a web page containing further class information.
![image](https://github.com/user-attachments/assets/3704d489-6c59-45f2-80d2-a991f17af141)

**Reply by iplaw67** [reply]
Could these web pages be stored in the instance directory structure, e.g. /docs ?

I'd have to find some way of enabling editing to update them, but I guess that might be a poor man's workaround to the limitation of not being able to store the data directly in a field of a corresponding table in the database, but just not a very efficient one ?

**Reply by jjdejong** [reply]
They sure could, but that is a quick and dirty way. I believe you need to put the folder in the `/public` sub-folder, so that you can access it with the URL https://<your site>/docs/file.html. Otherwise, I've foreseen the possibility of populating a `classifier_value` table with "long" text for this type of use, but I don't quite remember how it was supposed to work, since I haven't used this at all... The information is [here](https://github.com/jjdejong/phpip/wiki/Tables#classifier_value). It probably hasn't been implemented in the UI.

**Reply by iplaw67** [reply]
Yeah, I was just thinking, after the fact, that sticking the HTML files in a subfolder might bring with it a whole load of pain when using git to update the local repo...I don't really want to have to go through any commit shenanigans if I can avoid it. As it was, when I upgraded today, I got errors about composer.json, then had to specifically check it out, and then composer install failed due to missing php8.2 (for which the instructions provided helped), and then the "--optimize-autoloader --no-dev" instruction which failed due to a missing laraval/ui component. That was not a particularly easy or enjoyable upgrade path ;-)

I'll have a look at the classifier_value table and see if there might be some hope in that direction.

**Reply by iplaw67** [reply]
I might have found a workaround using the Keyword classifier, which you can "abuse" to insert TM G&S. Obviously, the display faces the problem of no wrap-around, but at least it allows me to have the G+S, which I can identify with a TMCL string in front. In the grand scheme of things, only a minor duplication of effort. I've not yet encountered a list of goods or services, that would exceed 64kb.

**Reply by iplaw67** [reply]
Just realized from your reply above that my suggestion above won't work if the Keyword classifier is also shared with all other members of a family.

**Comment by iplaw67** [comment]
Also related to trademark management
Perhaps I am missing some understanding as to how the UI works, but is there anyway to upload a representation of the mark, e.g. for figurative/semi-figurative marks and have that image displayed on the matter page / detail ? 
Currently, all I see as possibilities in the drop down list of addable fields are text value based entries, but no binary objects (such as image files).

For patents:
In a similar vein, it would be nice to be able to upload a representative figure of the invention and have it displayed on the detail page. Is this possible?
Thanks in advance for any pointers.

**Reply by jjdejong** [reply]
Sure, in the classifiers, you can add a picture. 
![image](https://github.com/user-attachments/assets/c385a792-b6c4-4516-b6cd-469b7787385c)

**Reply by iplaw67** [reply]
Ah, thanks !

**Comment by iplaw67** [comment]
A UI question now : is there any way to make the action buttons (e.g. "plus" button), and some of the UI elements, bigger ?
For example, even on a large 27" screen, the UI elements are quite small, and I've found that I have to manually increase the magnification of the page in Chrome to more than 125% in order to be able to work comfortably with the UI.

Is there some setting I can use in a config file, or a UI parameter, to make the display size adapt to the screen resolution I'm running ?

**Reply by jjdejong** [reply]
Well, that is the first time I hear that! The icons are sized so that they fit in their headers or line spacing. Making them bigger would break the display in many places. They are handled as fonts, so they follow the font sizing. I believe the size may depend on your navigator. My Firefox shows them like this, which is amply sufficient, since there are not many: 
![image](https://github.com/user-attachments/assets/72bcaa3a-f0ec-460f-94ec-84977a6fc761)

**Reply by iplaw67** [reply]
Must be a Chrome thing, on my screen they are tiny and I'm blind as a bat 🤣 
The screenshot below is at 100% magnification:
![Screenshot](https://github.com/user-attachments/assets/f75ec87d-a0c8-4d60-a3b7-e4667bb018dd)

**Reply by iplaw67** [reply]
Here's the add button in the Classifier dialog
![AddClassifier](https://github.com/user-attachments/assets/09959ac9-2b04-4b3e-abdd-2a5ab2bf9326)

**Reply by jjdejong** [reply]
The first one is just an indicator that shows you can click on the header bar. The second is indeed small, but there is only one button and you know what it's for, so there's no real issue, right? Plus you have the tool-tips when you hover over the buttons. The thing is that increasing the size of these icons breaks the line spacing, causing stuttering when showing and hiding the icons as you move the mouse.

**Comment by srdco** [comment]
Is there a way to store a DAS or US Confirmation Number in phpIP?

**Reply by jjdejong** [reply]
There are plenty of places you can store things attached to a matter, like in notes fields, classifiers (by creating new types if necessary), or events if you want to assign them a date (and you can create new event names if necessary).

**Comment by srdco** [comment]
Got it. Confirmation codes are used heavily in the US, so I was hoping 
to have a place they could be easily accessed using DOCX templates. Any 
pointers on how to extend to make that available?
(I've written in PHP & MYSQL, but have never used Laravel)

Thanks!

**Reply by jjdejong** [reply]
Put them in "Alt. Ref" available via the Advanced matter edition. This field is available for the templates.

Also, I've been thinking of adding a "Patent Office" actor role for adding patent offices to the matters, and then you could put the information in the reference of a "Patent Office" actor.

**Reply by jjdejong** [reply]
Note that adding a "Patent Office" role is mere configuration available to any user, by adding a new entry in the roles table. However, recovering the patent office reference for the templates requires adding the field to the ones made available in [DocumentMergeService.php](https://github.com/jjdejong/phpip/blob/master/app/Services/DocumentMergeService.php#37).

**Comment by iplaw67** [comment]
If I want to import a batch of cases into phpip via a multiple insert (either via a sql statement or via a CSV import over a suitable GUI), is it sufficient to just import the corresponding information into the event table to start with, or would I have to start somewhere else e.g. the matter table ?

**Reply by jjdejong** [reply]
This is complicated, because the data is in several linked tables. You need the matter information first. But if you want to import data that is already published, use the `Matters->Create family from OPS` function. This will create a whole family of patents from a single publication number using the EPO's db service "[OPS](https://github.com/jjdejong/phpip/wiki/Automatic-patent-family-import-from-Open-Patent-Services-(OPS))".

**Reply by iplaw67** [reply]
The OPS won't work for the trademark cases I need to import, unfortunately, but I'll find a workaround.

**Reply by jjdejong** [reply]
The import procedure is to first create the matters and actors. You can then link events and actors. Creating certain events will automatically create tasks according to the Rules table, such as renewals, office deadlines, etc. Basically you need to perform at least four batches of inserts, in the tables `matter`, `actor`, `event`, and `matter_actor_lnk`, the latter also assigning roles to the actors from the `actor_role` table.

I usually start from a .csv file including all the data, importing that in to a MySQL table, and then dispatch the data with a bunch of SQL queries. It's not a straightforward task and requires good knowledge of the tables structure.

**Comment by srdco** [comment]
What is the preferred way to record the national office's application number of a national stage application? For example, when I put the filing number in the "Entered" event, it still wants to use the PCT application number for the app number, which is a problem during prosecution because the USPTO doesn't refer to it that way.

**Reply by jjdejong** [reply]
We put it in the "Filed" event that has the PCT filing date.
![image](https://github.com/user-attachments/assets/2a66b3d7-4a2d-455c-b1b3-c1594c28598b)

**Reply by srdco** [reply]
OK -- so just overwrite the "filed" NUMBER (but not the date) when entering national stage

**Reply by jjdejong** [reply]
Absolutely! The "Parent filed" event is intended to show the PCT filing number, if you want it.

**Reply by papoteur-mga** [reply]
Hi @jjdejong 
Thus an improvement could be that, when creating the matter "National phase", the field would be populated with the PCT number.
What about the PCT publication number?

**Reply by jjdejong** [reply]
> Thus an improvement could be that, when creating the matter "National phase", the field would be populated with the PCT number.

Man, this is exactly what happens already.

> What about the PCT publication number?

The event is copied too.

The filing, publication and priority events of the PCT are copied in the national phases when using the "Nat. Phase" button. Moreover, a "Parent Filed" event is created pointing to the PCT filing for tracking the PCT information if you change the filing and publication numbers.

**Comment by iplaw67** [comment]
OPS error message
So, I thought I'd try out the OPS import function, but I get an error message.
I entered an EP publication number in the prescribed format.
I set my own case reference, as I don't want phpip to generate it for me. In this case, the caseref is "2385EP".
The Responsible person is my phpip user name.

The error message reads as follows :

Unexpected result:(
"message": "Trying to access array offset on null",
"exception": "Error exception",
"file": "/var/www/html/phpip/vendor/laravel/framework/src/Illuminate/Http/Client/Response.php",
"line": 418,
"trace": [
{
"file": "/var/www/html/phpip/vendor/laravel/framework/src/Illuminate/Foundation/Bootstrap/HandleExceptions.php",
"line": 290,
"function": "handleError",
"Class": "Illuminate\\Foundation\\Bootstrap\\HandleExceptions",
"type": "->"
},
{
"file": "/var/www/html/phpip/vendor/laravel/framework/src/Illuminate/Http/Client/Response.php",
"line": 418,
"function": "Illuminate\\Foundation\\Bootstrap\\(closure)",
"Class": "Illuminate\\Foundation\\Bootstrap\\HandleExceptions",
"type": "->"
},
{
...

It would be helpful if there was some way to copy this error message from somewhere, as the screen message isn't copiable (for me at least).

**Reply by iplaw67** [reply]
Ah found the rest in the laravel.log file in /storage/logs/, copy-pasted from first error, in attached txt file.
[ErrorOPSImport.txt](https://github.com/user-attachments/files/18811709/ErrorOPSImport.txt)

**Reply by jjdejong** [reply]
Did you create an OPS account and set the API key, etc., as indicated in the Wiki?

**Reply by iplaw67** [reply]
Oops, lol!

**Reply by jjdejong** [reply]
RTFM!

**Comment by iplaw67** [comment]
Espacenet URL construction
When clicking on a publication number in a matter to access the corresponding Espacenet page, is there any way to automatically differentiate between the A1 and the B1 URLs (or any other publication type, e.g. A3, A4, B2, C1, C2, etc) ?
Currently, if I add the publication type as part of the field in the matter, I get an error on the Espacenet  page when clicking the link formed in phpip.
Similarly, if I add the country code in front of the number (which is fairly standard practice in other IP management software that I've used in the past), I also get an error because the URL that is formed doesn't correspond to what Espacenet is expecting.

**Reply by jjdejong** [reply]
I haven't investigated, didn't really need that functionality. Why don't you?

**Reply by iplaw67** [reply]
When I get time after setting everything up, I might well do that ;-)
I don't strictly need it either, but as it was available, I thought I'd check it out.

**Reply by jjdejong** [reply]
I didn't directly answer these:

> Currently, if I add the publication type as part of the field in the matter, I get an error on the Espacenet page when clicking the link formed in phpip.

That's because the publication type is not used in the query link that expects a plain number for the "NR" parameter. 

> Similarly, if I add the country code in front of the number (which is fairly standard practice in other IP management software that I've used in the past), I also get an error because the URL that is formed doesn't correspond to what Espacenet is expecting.

Same thing. The query link expects a country code to be provided separately in parameter "CC", which code is simply taken from the matter country code. We don't recommend including the country code in the numbers, because that information is already present elsewhere in each matter (and this of course doesn't work as is with the Espacenet query system, unless a number pre-processing is developed that extracts a potential country code prefix).

The numbers are processed differently depending on their type (application, publication, grant) and on the specific Espacenet function used (bibliography, register...), so the best way to deal with this is to process the "components" of a "number" separately.

**Comment by iplaw67** [comment]
Line wrapping in the Classifier window
Is there any way to configure line wrapping of text portions within a window as drawn on screen, or implement a vertical scrollbar ?
For example, if I copy/paste the abstract of a patent application in the Classifier section, not only does it exceed the size of the outline border for the Abstract entry, but it also only displays a horizontal scrollbar, which is neither particularly graceful from a UI perspective, nor particularly useful from a user perspective.

**Reply by jjdejong** [reply]
The classifiers weren't really designed for large amounts of text, especially text available elsewhere. Why do you want to duplicate the information available one click away on Espacenet? Further development would be required to adapt the UI.

**Reply by iplaw67** [reply]
Personally, I can do without having the abstract, but I saw it there in the list of possibilities, so I tried it out, and it didn't display as I expected.

**Reply by jjdejong** [reply]
Right, the "Abstract" is available in the list of classifier types... There's a bunch of items in there that were added as potential right at the beginning, but never explored by us... Putting large texts in there would require some UI tweaking, like displaying a popup with the text.

**Comment by iplaw67** [comment]
National EP validation question
How does one deal with the situation in which we are asked to become the national representatives (with or without renewal responsibility) of a granted EP ?
Is the procedure to first create the EP case (which might depend on a prior PCT application), and then create the national validation from that ?
Alternatively, is it possible to just create a single new FR/EP case ?
I'm trying to understand whether I would have to potentially create 2 or 3 cases, or whether I can just create a single one corresponding to the specific circumstance.

**Reply by jjdejong** [reply]
You can create from the blue a new case as a PCT national phase or an EP validation. We like to have the parent cases in our system, so we would import the family from OPS and delete the cases we're not handling.

**Reply by iplaw67** [reply]
Thanks !

**Comment by iplaw67** [comment]
IDS list generation
Does phpip have any functionality to manage IDS or generate a list of documents with links to provide to US counsel ?

**Reply by jjdejong** [reply]
No. phpIP does not have a lot of functionality specific to a few exotic countries. The functionality remains generic or specific to the needs of the developers, who are European practitioners.

What's happening? There are two US practitioners that have decided at the same time to crash test phpIP and ask for features. Guys, if you don't find the features you're looking for, why don't you contribute them, eh?

**Reply by srdco** [reply]
@iplaw67 - You might look at SyncIDS as a commercial service to provide this functionality (as well as auto-filing)

**Comment by srdco** [comment]
Well, I haven't asked for new features -- just how things are supposed to work. I'm trying to find the limit of what's intended so I can make sure I'm using it 'properly.' Any US specific features I'm well aware I need to add or provide elsewhere. I fully understand this is a FOSS project. I've been shocked (pleasantly) by the help and support.

**Comment by srdco** [comment]
Quick Q -- where should I look for how to integrate Renewr with phpIP? I'm not seeing it in the Wiki, but I'm sure I could be missing something. Renewr didn't seem sure on their side. I'm happy to try to jot down some instructions from my experience if someone can point me in the right direction to figuring it out.

**Reply by papoteur-mga** [reply]
> Quick Q -- where should I look for how to integrate Renewr with phpIP? I'm not seeing it in the Wiki, but I'm sure I could be missing something. Renewr didn't seem sure on their side. I'm happy to try to jot down some instructions from my experience if someone can point me in the right direction to figuring it out.

Hello,
You can see at first here:
https://github.com/jjdejong/phpip/commit/dcfcc08b5e81b915e1d742af3b804a03a3520528
